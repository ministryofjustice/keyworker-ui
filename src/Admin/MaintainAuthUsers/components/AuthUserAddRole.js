import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import Select from '@govuk-react/select'
import Button from '@govuk-react/button'
import GridRow from '@govuk-react/grid-row'
import GridCol from '@govuk-react/grid-col'
import { BLACK, GREY_3 } from 'govuk-colours'
import { SPACING } from '@govuk-react/constants'
import { authRoleListType, errorType } from '../../../types'
import { lookupMeta } from '../../../govuk-helpers'

export const ButtonContainer = styled('div')`
  display: flex;
  flex-direction: row;
  button {
    margin-right: ${SPACING.SCALE_2};
  }
`

const AuthUserAddRole = ({ roleFilterList, handleRoleAddChange, roleList, handleAdd, handleCancel, error }) => {
  const roleListWithoutCurrentRoles = roleFilterList.filter(
    filteredRole => !roleList.some(currentRole => currentRole.roleCode === filteredRole.roleCode)
  )

  const roleListOptions = roleListWithoutCurrentRoles
    ? roleListWithoutCurrentRoles.map(role => (
        <option key={`role_option_${role.roleCode}`} data-qa={`${role.roleCode}_option`} value={role.roleCode}>
          {role.roleName}
        </option>
      ))
    : []

  const rolesAvailable = roleListOptions && roleListOptions.length > 0

  return (
    <GridRow>
      <GridCol setWidth="two-thirds">
        <form onSubmit={handleAdd}>
          {!rolesAvailable && <div data-qa="no-roles">No roles available</div>}
          {rolesAvailable && (
            <Select
              label="Choose new role"
              mb={6}
              meta={lookupMeta('role', error)}
              htmlFor="role"
              input={{ 'data-qa': 'role', name: 'role', onChange: handleRoleAddChange }}
            >
              <option key="choose" value="--">
                -- Select --
              </option>
              {roleListOptions}
            </Select>
          )}
          <ButtonContainer>
            {rolesAvailable && (
              <Button type="submit" data-qa="add-button">
                Add role
              </Button>
            )}
            <Button
              type="button"
              data-qa="cancel-button"
              onClick={handleCancel}
              buttonColour={GREY_3}
              buttonTextColour={BLACK}
            >
              Cancel
            </Button>
          </ButtonContainer>
        </form>
      </GridCol>
    </GridRow>
  )
}

AuthUserAddRole.propTypes = {
  handleRoleAddChange: PropTypes.func.isRequired,
  handleAdd: PropTypes.func.isRequired,
  handleCancel: PropTypes.func.isRequired,
  roleFilterList: authRoleListType.isRequired,
  roleList: authRoleListType.isRequired,
  error: errorType.isRequired,
}

export default AuthUserAddRole
