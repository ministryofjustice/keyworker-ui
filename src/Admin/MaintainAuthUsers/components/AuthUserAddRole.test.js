import React from 'react'
import renderer from 'react-test-renderer'
import Label from '@govuk-react/label'
import AuthUserAddRole from './AuthUserAddRole'

const roles = [
  { roleCode: 'ROLE_1', roleName: 'Role 1' },
  { roleCode: 'ROLE_2', roleName: 'Role 2' },
  { roleCode: 'ROLE_3', roleName: 'Role 3' },
]

const stubFunc = () => {}

describe('Auth user add roles', () => {
  describe('no roles', () => {
    const wrapper = renderer.create(
      <AuthUserAddRole
        roleList={roles}
        roleFilterList={[]}
        handleRoleAddChange={stubFunc}
        handleAdd={stubFunc}
        handleCancel={stubFunc}
        error=""
      />
    )
    it('should render correctly with user and no roles', () => {
      expect(wrapper.toJSON()).toMatchSnapshot()
    })

    it('should display message about no roles to add', () => {
      const noRowsMessage = wrapper.root.findByProps({ 'data-qa': 'no-roles' })
      // last table row should have message
      expect(noRowsMessage.props.children).toEqual('No roles available')
    })
  })

  it('should render the add role page with a filtered list of roles', () => {
    const component = renderer.create(
      <AuthUserAddRole
        roleList={[{ roleCode: 'ROLE_2', roleName: 'Role 2' }]}
        roleFilterList={roles}
        handleAdd={jest.fn()}
        handleCancel={jest.fn()}
        handleRoleAddChange={jest.fn()}
        error=""
      />
    )
    expect(component.toJSON()).toMatchSnapshot()
  })

  describe('errors', () => {
    const component = renderer.create(
      <AuthUserAddRole
        roleList={[{ roleCode: 'ROLE_2', roleName: 'Role 2' }]}
        roleFilterList={roles}
        handleAdd={jest.fn()}
        handleCancel={jest.fn()}
        handleRoleAddChange={jest.fn()}
        error={[{ targetName: 'role', text: 'something went wrong' }]}
      />
    )
    it('should render an error on the select if error occurred', () => {
      expect(component.toJSON()).toMatchSnapshot()
    })
    it('should display error', () => {
      expect(component.root.findByType(Label).props.error).toEqual('something went wrong')
    })
  })
})
