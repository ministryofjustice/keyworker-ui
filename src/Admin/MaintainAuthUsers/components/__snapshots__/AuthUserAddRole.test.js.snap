// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth user add roles errors should render an error on the select if error occurred 1`] = `
<div>
  <label
    className="sc-bdVaJa gDsdZv"
    htmlFor="role"
  >
    <span
      className="sc-bwzfXH lbxcXc"
    >
      Choose new role
    </span>
    <span
      className="sc-htpNat lbiKAP"
    >
      something went wrong
    </span>
    <select
      className="sc-ifAKCX iaVBIO"
      id="role"
      name="role"
      onChange={[MockFunction]}
    >
      <option
        value="--"
      >
        -- Select --
      </option>
      <option
        id="ROLE_1_option"
        value="ROLE_1"
      >
        Role 1
      </option>
      <option
        id="ROLE_3_option"
        value="ROLE_3"
      >
        Role 3
      </option>
    </select>
  </label>
  <button
    className="sc-EHOje hEaLol"
    disabled={false}
    id="add-button"
    onClick={[MockFunction]}
  >
    Add role
  </button>
</div>
`;

exports[`Auth user add roles no roles should render correctly with user and no roles 1`] = `
<div>
  <div
    id="no-roles"
  >
    No roles available
  </div>
</div>
`;

exports[`Auth user add roles should render the add role page with a filtered list of roles 1`] = `
<div>
  <label
    className="sc-bdVaJa cmpKXw"
    htmlFor="role"
  >
    <span
      className="sc-bwzfXH lbxcXc"
    >
      Choose new role
    </span>
    <select
      className="sc-ifAKCX kDzsBS"
      id="role"
      name="role"
      onChange={[MockFunction]}
    >
      <option
        value="--"
      >
        -- Select --
      </option>
      <option
        id="ROLE_1_option"
        value="ROLE_1"
      >
        Role 1
      </option>
      <option
        id="ROLE_3_option"
        value="ROLE_3"
      >
        Role 3
      </option>
    </select>
  </label>
  <button
    className="sc-EHOje hEaLol"
    disabled={false}
    id="add-button"
    onClick={[MockFunction]}
  >
    Add role
  </button>
</div>
`;
